src/util/taproot.rs

```diff
@@ -82,12 +92,753 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+            a.hash
+                .consensus_encode(&mut eng)
+                .expect("Engines don't err");
+        };
+        Ok(Self {
+            hash: sha256::Hash::from_engine(eng),
+            leaves: all_leaves,
+        })
+    }
+}
+
+// Internally used structure to store information about taproot leaf node
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct LeafInfo {
+    /// The underlying script
+    script: Script,
+    /// The leaf version
+    ver: LeafVersion,
+    /// The merkle proof(hashing partners) to get this node
+    merkle_branch: TaprootMerkleBranch,
+}
+
+impl LeafInfo {
+    // Create an instance of Self from Script with default version and no merkle branch
+    fn new(script: Script, ver: LeafVersion) -> Self {
+        Self {
+            script: script,
+            ver: ver,
+            merkle_branch: TaprootMerkleBranch(vec![]),
+        }
+    }
+
+    // Compute a leaf hash for the given leaf
+    fn hash(&self) -> sha256::Hash {
+        compute_leaf_hash(&self.script, self.ver)
+    }
+}
+
+/// function to compute leaf hash from components
+pub fn compute_leaf_hash(script: &Script, ver: LeafVersion) -> sha256::Hash {
+    let mut eng = TapLeafHash::engine();
+    ver.as_u8()
+        .consensus_encode(&mut eng)
+        .expect("engines don't err");
+    script
+        .consensus_encode(&mut eng)
+        .expect("engines don't err");
+    sha256::Hash::from_engine(eng)
+}
+
+/// helper function to obtain tap tweak
+pub fn compute_tap_tweak(
+    merkle_root: Option<sha256::Hash>,
+    internal_key: schnorr::PublicKey,
+) -> sha256::Hash {
+    let mut eng = TapTweakHash::engine();
+    // always hash the key
+    eng.input(&internal_key.serialize());
+    if let Some(h) = merkle_root {
+        eng.input(&h);
+    } else {
+        // nothing to hash
+    }
+    sha256::Hash::from_engine(eng)
+}
+
+/// The Merkle proof for inclusion of a tree in a taptree hash
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct TaprootMerkleBranch(Vec<sha256::Hash>);
+
+impl TaprootMerkleBranch {
+    /// Obtain a reference to inner
+    pub fn as_inner(&self) -> &[sha256::Hash] {
+        &self.0
+    }
+
+    /// Create a merkle proof from slice
+    pub fn from_slice(sl: &[u8]) -> Result<Self, TaprootError> {
+        if sl.len() % TAPROOT_CONTROL_NODE_SIZE != 0 {
+            Err(TaprootError::InvalidMerkleBranchSize(sl.len()))
+        } else if sl.len() > TAPROOT_CONTROL_NODE_SIZE * TAPROOT_CONTROL_MAX_NODE_COUNT {
+            Err(TaprootError::InvalidMerkleTreeDepth(
+                sl.len() / TAPROOT_CONTROL_NODE_SIZE,
+            ))
+        } else {
+            let mut inner = Vec::with_capacity(sl.len() / TAPROOT_CONTROL_NODE_SIZE);
+            let mut start = 0;
+            while start < sl.len() {
+                let elem = sha256::Hash::from_slice(&sl[start..start + TAPROOT_CONTROL_NODE_SIZE])
+                    .expect("must be multiple for 32");
```

Comment:
This condition is checked at the start of the function. `sl.len() % TAPROOT_CONTROL_NODE_SIZE != 0`, so this panic can never be triggered. 

I would also need to create a new error type for propagating errors that can never occur and could be confusing to the reader.

Reply:
> and other places

If you find any reachable panic in this PR, that is a bug.

Reply:
> I would also need to create a new error type for propagating errors that can never occur and could be confusing to the reader.

That makes sense. The large number of `expect` calls just looked odd on first review. Will keep this in mind on further review.

Reply:
Perhaps instead of "must be multiple for 32" we could use "unreachable: we check this above" or something like that. Another option is to have a comment `// Unreachable: checked above.`  - I'd definitely prefer to see some explicit thing so ever dev that reads theses `expect` calls does not have to waste clock cycles wondering why we panic.